### DISCUSSION TITLE
Qilin as a library for Analysis Android code

### DISCUSSION THREAD

**Discussion by Alireza-Ardalani** [QilinPTA/Qilin#9] [Q&A]
Hi,

Is there any way to use Qilin as library to analysis android (.apk file) and get k-CFA call graph?

I appreciate fro any help or guideline.

Thanks!

**Comment by DongjieHe** [comment]
Hi, currently there is no directly way to use Qilin to analyze Android apps. However, you can enable this functionality with minor modifications. In this issue (https://github.com/QilinPTA/Qilin/issues/8), I explained how to do that. So please refer to the issue first.

If there are any questions, welcome to further discussion.

**Reply by Alireza-Ardalani** [reply]
Hi, I try to analyse android  (.apk) using Qilin, it was successful, but now I realized that, I do not have some method from android.*
in methodContext list or other place. In fact, my problem is here:

LocationManager locationManager = (LocationManager) getSystemService(LOCATION_SERVICE);
Location location = locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);

I cannot find "getLastKnownLocation" method, which I checked it is not Fantom, and actually I can track it till here:

Solver.processStmts() --> recordCallStmts(momc, mpag.getInvokeStmts());

And: in method --> Solver.recordCallStmts()

I check that and statement for "getLastKnownLocation" added to:
virtualCallSiteQueue.add(virtualCallSite);

$r4 = virtualinvoke r3.<android.location.LocationManager: android.location.Location getLastKnownLocation(java.lang.String)>("gps")
CallSite: $r4 = virtualinvoke r3.<android.location.LocationManager: android.location.Location getLastKnownLocation(java.lang.String)>("gps")


I lost track of the issue beyond this point. If anything comes to mind regarding what might be happening and how I can handle it, I would really appreciate your insights.

**Comment by DongjieHe** [comment]
Hi, I guess the points-to-set of "r3" is empty. So even you could obverse the call statement being processed during the pointer analysis, "getLastKnownLocation" method is never reachable.

I'm guessing the call to "getSystemService" is related to any native code that is not handled by Qilin. If this is in the case, you should manually handle the native code by yourself.
